libp2p/host.ss

```diff
@@ -1,58 +1,60 @@
-;;; -*- Gerbil -*-
+;;; -*-
 ;;; libp2p host implementation
 
 (import :std/io
         :std/misc/timeout
         :std/format
+        :std/generic
         :std/sugar
         :std/crypto
         :std/text/base58
+        :std/text/utf8
         :std/logger
         "utils"
-        "multiaddress")
+        "multiaddress"
+        (prefix-in "protocols/multistream-select/lib" mux#))
 
+(start-logger! (current-output-port))
+
+;;~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+;; Peer Info
+;;~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+(defstruct peer-info (peer-id multi-address public-key))
 ;; TODO This should derive the peer-id from the key-pair
 (def (make-peer-id key)
   (base58-encode key))
 
-(defstruct peer-info (peer-id multi-address public-key))
-
+;;~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+;; Host Interface
+;;~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+;; TODO I should hide [id, key, multi-address] inside of the peer store.
 (defstruct p2p-host (id key multi-address listener peer-store handlers connections)
   constructor: init!)
 
 (defmethod {init! p2p-host}
   (lambda (self (key (keygen/ed25519)) (multi-addr "ip4/127.0.0.1/tcp/4444"))
-    (set! (@ self multi-address) (multi-address multi-addr))
-    (set! (@ self peer-store) (make-hash-table))
-    (set! (@ self handlers) (make-hash-table))
-    (set! (@ self connections) (make-hash-table))
+    (set! self.multi-address (multi-address multi-addr))
+    (set! self.peer-store (make-hash-table))
+    (set! self.handlers (make-hash-table))
+    (set! self.connections (make-hash-table))
     ;; TODO Make this a proper implementation
-    (set! (@ self id) (make-peer-id key))))
-
-(def (tcp-listener address (proc 123))
-  (let* ((addr (get-tcp-addr address))
-         (sock (tcp-listen addr)))
-    (while #t
-      (try
-       (let (cli (ServerSocket-accept sock))
-         (when cli
-           (println  (format "~a" cli))
-           (println (format "Accepted connection from ~a" (StreamSocket-peer-address cli)))
-           ;;(spawn proc cli)
-           ))
-       (catch (e)
-         (errorf "Error accepting connection: ~a" e))))))
+    (set! self.id (make-peer-id key))))
 
+;; TODO maybe make this dispatch to a generic?
 (defmethod {listen! p2p-host}
   (lambda (self)
-    (def addr (@ self multi-address))
+    (def addr self.multi-address)
     (match addr
-      ((? tcp-ip?) (set! (@ self listener)
-                     (spawn tcp-listener addr))))))
+      ((? tcp-ip?) (set! self.listener
+                     (spawn tcp-listener addr self.handlers))))))
 
 (defmethod {set-protocol-handler! p2p-host}
   (lambda (self protocol handler)
-    (hash-put! (@ self handlers) protocol handler)))
+    (hash-put! self.handlers protocol handler)))
+
+(defgeneric connect
+  (lambda args #f))
```

Comment:
i am not convinced this should be a generic, all you take is a multiaddr, and dispatch based on this.

Much better to implement as a method in the host, which maintains a mapping of (multiaddr) proto -> dialer